centreon/src/Core/Service/Application/UseCase/FindServices/FindServices.php

```diff
@@ -0,0 +1,180 @@
+        private readonly ReadServiceRepositoryInterface $readServiceRepository,
+        private readonly ReadHostRepositoryInterface $readHostRepository,
+        private readonly ReadServiceCategoryRepositoryInterface $readServiceCategoryRepository,
+        private readonly ReadServiceGroupRepositoryInterface $readServiceGroupRepository,
+        private readonly RequestParametersInterface $requestParameters,
+        private readonly ContactInterface $user
+    ) {
+    }
+
+    /**
+     * @param FindServicesPresenterInterface $presenter
+     */
+    public function __invoke(FindServicesPresenterInterface $presenter): void
+    {
+        try {
+            if (
+                ! $this->user->hasTopologyRole(Contact::ROLE_CONFIGURATION_SERVICES_READ)
+                && ! $this->user->hasTopologyRole(Contact::ROLE_CONFIGURATION_SERVICES_WRITE)
+            ) {
+                $this->error(
+                    "User doesn't have sufficient rights to see services",
+                    ['user_id' => $this->user->getId()]
+                );
+                $presenter->presentResponse(
+                    new ForbiddenResponse(ServiceException::accessNotAllowed())
+                );
+
+                return;
+            }
+
+            if ($this->user->isAdmin()) {
+                $services = $this->readServiceRepository->findByRequestParameter($this->requestParameters);
+            } else {
+                $accessGroups = $this->readAccessGroupRepository->findByContact($this->user);
+                $services = $this->readServiceRepository->findByRequestParameterAndAccessGroup($this->requestParameters, $accessGroups);
+            }
+
+            $presenter->presentResponse($this->createResponse($services));
+        } catch (RequestParametersTranslatorException $ex) {
+            $presenter->presentResponse(new ErrorResponse($ex->getMessage()));
+            $this->error($ex->getMessage(), ['trace' => $ex->getTraceAsString()]);
+        } catch (\Throwable $ex) {
+            $presenter->presentResponse(
+                new ErrorResponse(ServiceException::errorWhileSearching($ex))
+            );
+            $this->error($ex->getMessage(), ['trace' => $ex->getTraceAsString()]);
+        }
+    }
+
+    /**
+     * @param ServiceLight[] $services
+     *
+     * @return FindServicesResponse
+     */
+    private function createResponse(array $services): FindServicesResponse
+    {
+        $categoryIds = [];
+        $groupIds = [];
+        $hostIds = [];
+        foreach ($services as $service) {
+            $categoryIds = array_merge($categoryIds, $service->getCategoryIds());
+            $hostIds = array_merge(
+                $hostIds,
+                $service->getHostIds(),
+                ServiceGroupRelation::getHostIds($service->getGroups())
+            );
+            $groupIds = array_merge($groupIds, ServiceGroupRelation::getServiceGroupIds($service->getGroups()));
+        }
```

Comment:
`array_merge` is very heavy. 
For those loops, we should avoid calling it several times if we can.
The naming of the variables may need to be changed in my suggestion 👇🏻 
This is basically a tradeoff memory/cpu.
```suggestion
        foreach ($services as $service) {
            $categoryIds[] = $service->getCategoryIds();
            $hostIds[] = $service->getHostIds();
            $hostIds[] = ServiceGroupRelation::getHostIds($service->getGroups());
            $groupIds[] = ServiceGroupRelation::getServiceGroupIds($service->getGroups());
        }
        $categoryIds = array_merge(...$categoryIds);
        $hostIds = array_merge(...$hostIds);
        $groupIds = array_merge(...$groupIds);
```